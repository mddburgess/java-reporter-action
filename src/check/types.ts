export enum RunCondition {
  disabled,
  autodetect,
  expected,
  required,
}
