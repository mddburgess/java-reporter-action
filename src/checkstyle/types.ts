import { relativePath } from "../common/utils";
import { AnnotationLevel } from "../github/types";
import LintAnnotation from "../results/LintAnnotation";

export default interface CheckstyleReport {
  violations: CheckstyleViolation[];
}

export interface CheckstyleViolation {
  filePath: string;
  line: number;
  column: number;
  rule: string;
  severity: CheckstyleSeverity;
  message: string;
}

export type CheckstyleSeverity = "error" | "warning" | "info";

export const toLintAnnotation = (violation: CheckstyleViolation): LintAnnotation => ({
  path: relativePath(violation.filePath),
  line: violation.line,
  level: getLevel(violation.severity),
  category: getCategory(violation.rule),
  type: getType(violation.rule),
  message: violation.message,
});

const getLevel = (severity: CheckstyleSeverity): AnnotationLevel => {
  switch (severity) {
    case "error":
      return "failure";
    case "warning":
      return "warning";
    case "info":
      return "notice";
    default:
      throw Error("Unexpected severity");
  }
};

const getCategory = (rule: string): string => {
  const idx = rule.lastIndexOf(".");
  return idx === -1 ? "" : rule.slice(0, idx);
};

const getType = (rule: string): string => {
  return rule.split(".").slice(-1)[0];
};
