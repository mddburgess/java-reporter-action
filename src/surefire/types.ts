export default interface SurefireReport {
  name: string;
  tests: number;
  passed: number;
  failures: number;
  errors: number;
  skipped: number;
  testCases: SurefireTestCase[];
}

export interface SurefireTestCase {
  className: string;
  testName: string;
  result?: SurefireTestResult;
  message?: string;
  stackTrace?: string;
}

export type SurefireTestResult = "success" | "failure" | "error" | "skipped";
