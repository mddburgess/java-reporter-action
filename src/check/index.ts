import * as core from "@actions/core";
import * as glob from "@actions/glob";
import NoReportsResult from "../common/no-reports";
import CheckRun from "../github/check-run";
import CheckResult from "./result";
import { RunCondition } from "./types";

export default abstract class Check<T> {
  private readonly type: string;
  private readonly friendlyName: string;
  readonly runCondition: RunCondition;
  private readonly checkRun: CheckRun;

  protected constructor(type: string, friendlyName: string) {
    this.type = type;
    this.friendlyName = friendlyName;
    this.runCondition = this.resolveRunCondition();
    this.checkRun = new CheckRun(this.type);
  }

  private resolveRunCondition() {
    const condition = core.getInput(this.type);
    switch (condition) {
      case "disabled":
      case "autodetect":
      case "expected":
      case "required":
        return RunCondition[condition];
      default:
        core.warning(`Invalid input: ${this.type} -- defaulting to autodetect`);
        return RunCondition.autodetect;
    }
  }

  async run() {
    if (this.runCondition === RunCondition.disabled) {
      core.warning(`${this.friendlyName} check is disabled.`);
      return;
    }
    if (this.runCondition >= RunCondition.expected) {
      await this.checkRun.queue();
    }
    const result = await this.runCheck();
    if (result.shouldCompleteCheck()) {
      await this.checkRun.complete(result);
    }
    core.info(`${this.friendlyName} check finished.`);
  }

  async runCheck() {
    const searchPaths = this.resolveSearchPaths();

    const reportPaths = await this.resolveReportPaths(searchPaths);
    if (reportPaths.length === 0) {
      return new NoReportsResult(this.friendlyName, this.runCondition, searchPaths);
    }

    const reports = this.readReports(reportPaths);
    return this.getResult(reports);
  }

  private resolveSearchPaths() {
    return core.getInput(`${this.type}-report-paths`, { required: true }).split(",");
  }

  private async resolveReportPaths(searchPaths: string[]): Promise<string[]> {
    core.startGroup(`Searching for ${this.friendlyName} reports`);
    searchPaths.forEach((searchPath) => core.info(searchPath));
    core.endGroup();

    const globber = await glob.create(searchPaths.join("\n"));
    const reportPaths = await globber.glob();

    core.startGroup(`Found ${reportPaths.length} ${this.friendlyName} reports`);
    reportPaths.forEach((reportPath) => core.info(reportPath));
    core.endGroup();

    return reportPaths;
  }

  private readReports(reportPaths: string[]) {
    return reportPaths
      .map((reportPath) => this.readReport(reportPath))
      .filter((report) => report !== undefined) as T[];
  }

  protected abstract readReport(reportPath: string): T | undefined;
  protected abstract getResult(reports: T[]): CheckResult;
}
