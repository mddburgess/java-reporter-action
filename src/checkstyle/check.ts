import Check from "../check";
import CheckstyleReport from "./types";
import CheckstyleParser from "./parser";
import CheckResult from "../check/result";
import CheckstyleResult from "./result";

export default class CheckstyleCheck extends Check<CheckstyleReport> {
  constructor() {
    super("checkstyle", "Checkstyle");
  }

  protected readReport(reportPath: string): CheckstyleReport | undefined {
    return new CheckstyleParser(reportPath).read();
  }

  protected getResult(reports: CheckstyleReport[]): CheckResult {
    return new CheckstyleResult(this.runCondition, reports);
  }
}
