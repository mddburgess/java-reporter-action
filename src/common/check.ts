import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import path from 'path';
import {getInput} from '@actions/core';

abstract class Check {

    private static readonly workspacePath = process.env.GITHUB_WORKSPACE || '';

    private readonly reportType: string;
    private readonly checkCondition: CheckCondition;
    private checkRunId?: number;

    protected constructor(reportType: string) {
        this.reportType = reportType;
        this.checkCondition = this.resolveCheckCondition();
    }

    private resolveCheckCondition() {
        const condition = getInput(this.reportType);
        switch (condition) {
            case 'required':
            case 'expected':
            case 'disabled':
                return CheckCondition[condition];
            case undefined:
                return CheckCondition.autodetect;
            default:
                core.warning(`Input '${this.reportType}' is invalid - must be one of ['required','expected','disabled']. Defaulting to 'expected'.`);
                return CheckCondition.expected;
        }
    }

    public async run() {
        const reportPaths = await this.findReports();

        const token = core.getInput('github-token', {required: true});
        const octokit = github.getOctokit(token);

        if (reportPaths.length > 0 || this.checkCondition >= CheckCondition.expected) {
            const response = await octokit.checks.create({
                ...github.context.repo,
                name: this.reportType,
                head_sha: github.context.payload.pull_request
                    ? github.context.payload.pull_request.head.sha
                    : github.context.sha,
                status: 'completed',
                conclusion: reportPaths.length ? 'success' :
                    this.checkCondition === CheckCondition.required ? 'failure' : 'skipped'
            });
            this.checkRunId = response.data.id;
        }

        core.info(`${this.reportType} check finished.`);
    }

    private async findReports() {
        core.startGroup(`Searching for ${this.reportType} reports...`);

        const searchPaths = this.reportSearchPaths();
        searchPaths.forEach(searchPath => core.info(searchPath));
        core.endGroup();

        const globber = await glob.create(searchPaths.join('\n'));
        const reportPaths = await globber.glob();

        core.startGroup(`Found ${reportPaths.length} ${this.reportType} reports.`);
        reportPaths.forEach(reportPath => core.info(path.relative(Check.workspacePath, reportPath)));
        core.endGroup();

        return reportPaths;
    }

    protected abstract reportSearchPaths(): string[];
}

enum CheckCondition {
    disabled,
    autodetect,
    expected,
    required
}

export default Check;
