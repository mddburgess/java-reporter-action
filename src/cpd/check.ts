import Check from "../check";
import CpdReport from "./types";
import CheckResult from "../check/result";
import CpdResult from "./result";
import CpdParser from "./parser";

export default class CpdCheck extends Check<CpdReport> {
  constructor() {
    super("cpd", "CPD");
  }

  protected readReport(reportPath: string): CpdReport | undefined {
    return new CpdParser(reportPath).read();
  }

  protected getResult(reports: CpdReport[]): CheckResult {
    return new CpdResult(this.runCondition, reports);
  }
}
