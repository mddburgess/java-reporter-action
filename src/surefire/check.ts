import SurefireReport from "./types";
import Check from "../check";
import SurefireParser from "./parser";
import CheckResult from "../check/result";
import SurefireResult from "./result";

export default class SurefireCheck extends Check<SurefireReport> {
  public constructor(private readonly classpath: string[]) {
    super("surefire", "Surefire");
  }

  protected readReport(reportPath: string): SurefireReport | undefined {
    return new SurefireParser(reportPath).read();
  }

  protected getResult(reports: SurefireReport[]): CheckResult {
    return new SurefireResult(this.runCondition, reports, this.classpath);
  }
}
