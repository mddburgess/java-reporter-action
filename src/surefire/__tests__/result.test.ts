import { RunCondition } from "../../check/types";
import SurefireParser from "../parser";
import SurefireResult from "../result";

describe("SurefireResult", () => {
  it("can handle an empty Surefire report", () => {
    const report = {
      name: "com.example.ExampleTest",
      tests: 0,
      passed: 0,
      failures: 0,
      errors: 0,
      skipped: 0,
      testCases: [],
    };
    const result = new SurefireResult(RunCondition.autodetect, [report]);

    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("success");
    expect(result.title).toBe("0 tests passed");
    expect(result.summary).toBeDefined();
    expect(result.text).toBeDefined();
    expect(result.annotations).toStrictEqual([]);
  });

  it("can handle a real Surefire report", () => {
    const report = new SurefireParser("src/surefire/__tests__/__fixtures__/v3/junit5.xml").read();
    expect(report).toBeDefined();

    const result = new SurefireResult(RunCondition.autodetect, [report!]); // eslint-disable-line
    expect(result.shouldCompleteCheck()).toBe(true);
    expect(result.conclusion).toBe("failure");
    expect(result.title).toBe("6 tests failed");
    expect(result.summary).toMatchSnapshot();
    expect(result.text).toMatchSnapshot();
    expect(result.annotations).toMatchSnapshot();
  });
});
