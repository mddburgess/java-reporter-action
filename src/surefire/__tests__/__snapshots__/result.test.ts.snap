// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurefireResult can handle a real Surefire report 1`] = `
"|Tests run|11|
|:-|-:|
|Passed|1|
|Failures|4|
|Errors|2|
|Skipped|4|"
`;

exports[`SurefireResult can handle a real Surefire report 2`] = `
"|Class|Tests|Passed|Failures|Errors|Skipped|
|:-|-:|-:|-:|-:|-:|
|\`report.name\`|11|1|4|2|4|"
`;

exports[`SurefireResult can handle a real Surefire report 3`] = `
Array [
  Object {
    "annotation_level": "failure",
    "end_line": 28,
    "message": "org.opentest4j.AssertionFailedError: Assertion message ==> expected: <5> but was: <4>
	at org.example.JUnit5Test.testAssertionFailureWithMessage(JUnit5Test.java:28)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.AssertionFailedError: Assertion message ==> expected: <5> but was: <4>
	at org.example.JUnit5Test.testAssertionFailureWithMessage(JUnit5Test.java:28)",
    "start_line": 28,
    "title": "Test failure: JUnit5Test.testAssertionFailureWithMessage",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 43,
    "message": "java.lang.RuntimeException
	at org.example.JUnit5Test.testError(JUnit5Test.java:43)",
    "path": "org.example.JUnit5Test",
    "raw_details": "java.lang.RuntimeException
	at org.example.JUnit5Test.testError(JUnit5Test.java:43)",
    "start_line": 43,
    "title": "Test error: JUnit5Test.testError",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 33,
    "message": "org.opentest4j.AssertionFailedError: 
	at org.example.JUnit5Test.testFail(JUnit5Test.java:33)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.AssertionFailedError: 
	at org.example.JUnit5Test.testFail(JUnit5Test.java:33)",
    "start_line": 33,
    "title": "Test failure: JUnit5Test.testFail",
  },
  Object {
    "annotation_level": "notice",
    "end_line": 1,
    "message": "Disable message",
    "path": "org.example.JUnit5Test",
    "raw_details": undefined,
    "start_line": 1,
    "title": "Test skipped: JUnit5Test.testDisabledWithMessage",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 48,
    "message": "java.lang.RuntimeException: Error message
	at org.example.JUnit5Test.testErrorWithMessage(JUnit5Test.java:48)",
    "path": "org.example.JUnit5Test",
    "raw_details": "java.lang.RuntimeException: Error message
	at org.example.JUnit5Test.testErrorWithMessage(JUnit5Test.java:48)",
    "start_line": 48,
    "title": "Test error: JUnit5Test.testErrorWithMessage",
  },
  Object {
    "annotation_level": "notice",
    "end_line": 58,
    "message": "org.opentest4j.TestAbortedException: Assumption failed: Assumption message
	at org.example.JUnit5Test.testSkippedWithMessage(JUnit5Test.java:58)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.TestAbortedException: Assumption failed: Assumption message
	at org.example.JUnit5Test.testSkippedWithMessage(JUnit5Test.java:58)",
    "start_line": 58,
    "title": "Test skipped: JUnit5Test.testSkippedWithMessage",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 38,
    "message": "org.opentest4j.AssertionFailedError: Failure message
	at org.example.JUnit5Test.testFailWithMessage(JUnit5Test.java:38)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.AssertionFailedError: Failure message
	at org.example.JUnit5Test.testFailWithMessage(JUnit5Test.java:38)",
    "start_line": 38,
    "title": "Test failure: JUnit5Test.testFailWithMessage",
  },
  Object {
    "annotation_level": "notice",
    "end_line": 1,
    "message": "public void org.example.JUnit5Test.testDisabled() is @Disabled",
    "path": "org.example.JUnit5Test",
    "raw_details": undefined,
    "start_line": 1,
    "title": "Test skipped: JUnit5Test.testDisabled",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 22,
    "message": "org.opentest4j.AssertionFailedError: expected: <5> but was: <4>
	at org.example.JUnit5Test.testAssertionFailure(JUnit5Test.java:22)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.AssertionFailedError: expected: <5> but was: <4>
	at org.example.JUnit5Test.testAssertionFailure(JUnit5Test.java:22)",
    "start_line": 22,
    "title": "Test failure: JUnit5Test.testAssertionFailure",
  },
  Object {
    "annotation_level": "notice",
    "end_line": 53,
    "message": "org.opentest4j.TestAbortedException: Assumption failed: assumption is not true
	at org.example.JUnit5Test.testSkipped(JUnit5Test.java:53)",
    "path": "org.example.JUnit5Test",
    "raw_details": "org.opentest4j.TestAbortedException: Assumption failed: assumption is not true
	at org.example.JUnit5Test.testSkipped(JUnit5Test.java:53)",
    "start_line": 53,
    "title": "Test skipped: JUnit5Test.testSkipped",
  },
]
`;
